### PR TITLE
Model loading

### PR DESCRIPTION
**Pull request by tigranfah**
This PR does three important changes

* fix the grad accumulation loss calculation
Previously the loss was calculated on the last grad accumulation step only, not it is aggregated across all the grad accumulation steps
* change the dump dir to the h100
* fix the resume functionality in the loading and saving phase
Make sure that the CheckpointManager's states is a dict containing the model, optimizer and lr scheduler wrapper objects to perform save and load operations accurately
To further verify the functionality of the resume, the following experiment is conducted, where
1. a training run is started 150 steps
2. then it is resumed from 50th step and trains for 100 steps
3. run 2 is resume from 100 steps and trains for 50 steps

One gets the following picture after the experiment, suggesting that the implementation is correct

run 1
Screenshot 2024-11-26 at 12 46 05 PM
run 2
Screenshot 2024-11-26 at 12 46 13 PM
run 3
Screenshot 2024-11-26 at 12 46 20 PM

### CODE REVIEW

File: torchtitan/checkpoint.py [hunk 1]
```diff
@@ -473,15 +487,28 @@ def load(self, step: int = -1) -> bool:
 
         # We won't have optimizer states to load, if we are loading a seed checkpoint
         states = {"model": self.states["model"]} if step == 0 else self.states
+        # PyTorch bug: (pytorch/pytorch#138575)
```
Comment (philippguevorguian) [inline torchtitan/checkpoint.py:490]: This issue is closed: https://github.com/pytorch/pytorch/pull/138575, is the idea that we use this for compatibility with older pytorch versions?

**Review by philippguevorguian — APPROVED** [review]
Couple of clarifying questions added, but LGTM

File: train.py [hunk 2]
```diff
@@ -319,6 +322,7 @@ def loss_fn(pred, labels):
             # it issues a single all-reduce for all parameters at once for better performance
             float8_handler.precompute_float8_dynamic_scale_for_fsdp(model_parts)
 
+            loss /= job_config.training.gradient_accumulation_steps
```
[the remark this answers is not in the document]
Comment (tigranfah) [reply train.py]: I think I put it in the wrong place, this division should be reflected in the backward pass

**Comment by tigranfah** [comment]
Seems like the grad acc is not implemented correctly.

The final loss should be batch losses in the gradient accumulation steps averaged together.

This implementation at commit [af370a7](https://github.com/YerevaNN/YNNtitan/pull/32/commits/af370a721530e1aa6cf0690fc71eecfe7e87a4f0) does not average the batch level losses when performing the backward pass, this results in an incorrect implementation, which can be captured in the below experiment

Model: debug_llama model
Steps: 100
Peek lr: 8e-4

(blue batch_size=6, grad_acc_steps=8, green batch_size=1, grad_acc_steps=48)
Screenshot 2024-11-30 at 11 40 32 AM

The final losses are clearly different.

On the other hand when averaging the batch level losses before the backward pass [6aca1b5](https://github.com/YerevaNN/YNNtitan/pull/32/commits/6aca1b55cf30afb4af3299e5e282d38db85cdbb7) and repeating the above experiment

(blue batch_size=6, grad_acc_steps=8, green batch_size=1, grad_acc_steps=48)
Screenshot 2024-11-30 at 11 53 41 AM

At least on this specific case the performance matches.

**Comment by tigranfah** [comment]
Another important thing to note, to optimize further, we can do the backward pass outside of the grad accumulation loop, but in that case we have to recalculate the mfu accordingly, so may be it can be implemented in the future PRs.
